### PR TITLE
[Keil5] some extensions of possible options

### PR DESCRIPTION
**Pull request by ohagendorf**
- for the option extension please see in parallel the project_generator_definitions PR https://github.com/project-generator/project_generator_definitions/pull/70
- a minor change in a doc string format in project.py - only three """ are necessary

### CODE REVIEW

File: project_generator/tools/uvision.py [hunk 1]
```diff
@@ -215,6 +215,20 @@ def _uvproj_set_TargetArmAds(self, uvproj_dic, project_dic):
...
+        pro_def = ProGenDef(tool_name)
+        if not pro_def.is_supported(project_dic['target'].lower()):
+            raise RuntimeError("Target %s is not supported. Please add them to https://github.com/project-generator/project_generator_definitions" % project_dic['target'].lower())
+        mcu_def_dic = pro_def.get_tool_definition(project_dic['target'].lower())
+        if not mcu_def_dic:
+             raise RuntimeError(
+                "Target definitions were not found for %s. Please add them to https://github.com/project-generator/project_generator_definitions" % project_dic['target'].lower())
+        try:
+            uvproj_dic['ArmAdsMisc']['useUlib'] = mcu_def_dic['TargetOption']['uLib'][0]
+            uvproj_dic['ArmAdsMisc']['OptFeed'] = mcu_def_dic['TargetOption']['OptFeed'][0]
```
Comment (0xc0170) [inline project_generator/tools/uvision.py]: ``OptFeed`` what does it do? I commented on progen def, I would leave ``uLib`` out of definitions

**Comment by ohagendorf** [comment]
OptFeed is the Use Cross-Module Optimization flag.

When you thing, that this is a project specific option lnke uLib then it should be removed here and implemented in mbed.

**Comment by 0xc0170** [comment]
> When you thing, that this is a project specific option lnke uLib then it should be removed here and implemented in mbed.

Yes, definitely ! I would like to keep definitions only target specific (=re-usability in projects). This would mean that this PR is just about fixing the docstring.

**Comment by ohagendorf** [comment]
But what do you think about the RegisterFile?

That is mcu specific and not project specific. Without this extension it is configured to a Freescale Header file.

**Comment by 0xc0170** [comment]
``RegisterFile`` sounds good to be added, first progendef needs to get it (parser + mcu_def dictionary), then we can merge this one.

**Comment by ohagendorf** [comment]
Maybe we can share the work.
I'll can continue with the STM32 mcus, which are available as Discovery and Nucleo boards because we have most of them, during the next days and adding uvision5 options together with the RegisterFile options.

File: project_generator/tools/uvision.py [hunk 2]
```diff
@@ -294,6 +294,7 @@ def _set_target(self, expanded_dic, uvproj_dic, tool_name):
             uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['Cpu'] = mcu_def_dic['TargetOption']['Cpu'][0].encode('utf-8')
             uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['FlashDriverDll'] = str(mcu_def_dic['TargetOption']['FlashDriverDll'][0]).encode('utf-8')
             uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['SFDFile'] = mcu_def_dic['TargetOption']['SFDFile'][0]
+            uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['RegisterFile'] = mcu_def_dic['TargetOption']['RegisterFile'][0]
```
Comment (0xc0170) [inline project_generator/tools/uvision.py]: As I understand, registerfile was added just for uvision5? That means not all targets define registerfile. Which is ok, but the message below is not correct. TODO should be removed (as it's obsolete). Thus it should just continue.

**Comment by ohagendorf** [comment]
The RegisterFileOption is also available in mbed uvision template file and that seems to be in a uvision4 file format. But I don't have Keil4 so I can't fill the option with correct values. Maybe everywhere a 'null'?

I'm very sorry but I don't understand what you mean with message below.

**Comment by 0xc0170** [comment]
Remove the logging message plus todo comment, this is how it should be:
```
        try:
            uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['Vendor'] = mcu_def_dic['TargetOption']['Vendor'][0]
            uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['Cpu'] = mcu_def_dic['TargetOption']['Cpu'][0].encode('utf-8')
            uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['FlashDriverDll'] = str(mcu_def_dic['TargetOption']['FlashDriverDll'][0]).encode('utf-8')
            uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['SFDFile'] = mcu_def_dic['TargetOption']['SFDFile'][0]
            uvproj_dic['Project']['Targets']['Target']['TargetOption']['TargetCommonOption']['RegisterFile'] = mcu_def_dic['TargetOption']['RegisterFile'][0]
        except KeyError:
           pass
    
```

And we should create an issue, all uvision5/4 should contain all those values, so try/except should be removed later. Feel free to create an issue for this and reference there this PR.

**Comment by 0xc0170** [comment]
Thanks, looks good to me. I'll test locally

**Pull request merged**